### PR TITLE
feat(data-collection): Add RDS Health Dashboard module

### PR DESCRIPTION
**Pull request by anishkgemr**
### New: `module-rds.yaml`
  - Service-specific module for RDS operational data collection
  - Collects pending maintenance actions across linked accounts (multi-account, multi-region)
  - Cluster snapshot collection via CID inventory Lambda (scheduler disabled by default — pending upstream integration)
  - 10 resources, 753 lines, LINKED topology
  - Named per CID guidelines: `module-rds` (generic, per-service, no verdict words)
  
  ### Inventory Module Improvements (all 14 crawlers)
  - Switched crawler config to Trusted Advisor pattern:
    - `MergeNewColumns` — crawler adds columns, never removes
    - `InheritFromTable` — partitions inherit table schema (no per-partition type drift)
    - `UpdateBehavior: UPDATE_IN_DATABASE` + `DeleteBehavior: LOG`
  - Prevents type inference issues when optional fields are null/empty (e.g., `taglist`, `maxallocatedstorage`)
  - Benefits all inventory objects: EC2, EBS, VPC, EKS, Lambda, RDS, OpenSearch, ElastiCache, etc.
  
  ### Pre-defined RDS Table Fixes
  - `inventory_rds_db_instances_data`:
    - Fixed `taglist`: `array<string>` → `array<struct<Key:string,Value:string>>`
    - Added: `readreplicasourcedbinstanceidentifier`, `dbname`, `enginelifecyclesupport`, `databaseinsightsmode`
  - `inventory_rds_db_clusters_data`:
    - Added `classification:json` parameter (was missing — caused empty table after crawl)
    - Fixed `dbclustermembers`: `string` →
  `array<struct<DBInstanceIdentifier:string,IsClusterWriter:boolean,DBClusterParameterGroupStatus:string,PromotionTier:int>
  >`
    - Added: `taglist`, `globalclusteridentifier`
  - `inventory_rds_db_snapshots_data`:
    - Fixed `taglist`: `array<string>` → `array<struct<Key:string,Value:string>>`
  - Added `rds-cluster-snapshots` to inventory `sub_modules`
  
  ### Wiring (per CID module guidelines)
  - `deploy-data-collection.yaml`: `IncludeRDSModule` parameter + `DeployRDSModule` condition + `RDSModule` nested stack
  - `deploy-in-linked-account.yaml`: `IncludeRDSModulePolicy` condition + `RDSModulePolicy` (least-privilege) + trust for
  `${ResourcePrefix}rds-LambdaRole`
  - `deploy-data-read-permissions.yaml`: Pass-through (ParameterGroups, Labels, Parameter, nested stack, StackSet)
  - NOT in `deploy-in-management-account.yaml` (LINKED-only module)
  
  ### Naming Consistency (§3 compliance)
  | Touch-point | Token |
  |-------------|-------|
  | File | `module-rds.yaml` |
  | Parameter | `IncludeRDSModule` |
  | Condition | `DeployRDSModule` |
  | Logical ID | `RDSModule` |
  | IAM Policy | `RDSModulePolicy` |
  | Trust | `${ResourcePrefix}rds-LambdaRole` |
  
  ### Security
  - IAM least-privilege: only `rds:DescribePendingMaintenanceActions` + `ec2:DescribeRegions` + `sts:AssumeRole` + scoped
  S3
  - S3 ListBucket scoped to `rds-*` prefix
  - `RDSModulePolicy` gated by `IncludeRDSModulePolicy` condition — no permissions granted unless module enabled
  - cfn_nag suppressions documented for wildcard-required actions
  - Temp file cleanup in both success and error paths
  
  ### Testing
  - ✅ Edge case: 2 bare instances (no tags, no backups, no autoscaling) — all tables created, views work
  - ✅ All 7 RDS integration tests passing (`test_from_scratch.py`)
  - ✅ No impact on non-RDS inventory tables (AMI, VPC, Lambda, network interfaces all pass)
  - ✅ Stack UPDATE tested (columns added, not dropped)
  - ✅ Clusters table classification fix validated (was causing empty table)
  - ✅ Multi-account collection works (3 accounts: 519424981504, 299150411582, 165821488895)
  
  ### Dependencies
  - Dashboard PR: #1521 in cloud-intelligence-dashboards-framework repo (merge this first)
  
  ### Files Changed (6)
  | File | Changes |
  |------|---------|
  | `data-collection/deploy/module-rds.yaml` | NEW (753 lines) |
  | `data-collection/deploy/module-inventory.yaml` | Crawler config + RDS table schema fixes |
  | `data-collection/deploy/deploy-data-collection.yaml` | IncludeRDSModule wiring |
  | `data-collection/deploy/deploy-in-linked-account.yaml` | Trust + RDSModulePolicy + parameter |
  | `data-collection/deploy/deploy-data-read-permissions.yaml` | Pass-through |
  | `test/test_from_scratch.py` + `test/utils.py` | Updated tests |

### CODE REVIEW

File: data-collection/deploy/deploy-data-collection.yaml [hunk 1]
```diff
@@ -1624,6 +1633,33 @@ Resources:
             - !Sub "${AWS::Region}"
             - !Join [ '', !Split [ ' ', !Ref RegionsInScope  ] ] # remove spaces
 
+  RDSHealthModule:
```
Comment (yprikhodko) [inline data-collection/deploy/deploy-data-collection.yaml]: I'd suggest to give it more generic name like RDSModule as we may expand the scope of this module further - Description: Retrieves RDS operational and configuration data across the AWS organization.
RDSHealth reduces the scope

File: data-collection/deploy/module-rds-health.yaml [hunk 2]
```diff
@@ -0,0 +1,1914 @@
+AWSTemplateFormatVersion: '2010-09-09'
+Transform: 'AWS::LanguageExtensions'
+Description: 'RDS Health Dashboard - Data Collection v1.0 (CID Plugin)'
```
Comment (yprikhodko) [inline data-collection/deploy/module-rds-health.yaml]: Retrieves RDS operational and configuration data across the AWS organization

File: data-collection/deploy/module-rds-health.yaml [hunk 3]
```diff
@@ -0,0 +1,1914 @@
...
+                    s3:prefix:
+                      - 'rds-*'
+                      - 'inventory/inventory-rds-*'
+        - PolicyName: 'RDSDescribe'
+          PolicyDocument:
+            Version: '2012-10-17'
+            Statement:
+              - Effect: 'Allow'
+                Action:
+                  - 'rds:DescribeDBEngineVersions'
```
Comment (yprikhodko) [inline data-collection/deploy/module-rds-health.yaml]: we have a reference module https://github.com/aws-solutions-library-samples/cloud-intelligence-dashboards-data-collection/blob/main/data-collection/deploy/module-reference.yaml which collects DescribeDBEngineVersions and DescribeDBMajorEngineVersions data e.g. non instance specific global reference data. No need to duplicate it in this module

File: data-collection/deploy/module-rds-health.yaml [hunk 4]
```diff
@@ -0,0 +1,1914 @@
...
+                  - 'rds:DescribeDBInstances'
```
Comment (yprikhodko) [inline data-collection/deploy/module-rds-health.yaml]: DescribeDBInstances and DescribeDBClusters collected by inventory module

File: data-collection/deploy/module-rds-health.yaml [hunk 5]
```diff
@@ -0,0 +1,1914 @@
...
+          BUCKET_NAME: !Ref DestinationBucket
+    Metadata:
+      cfn_nag:
+        rules_to_suppress:
+          - id: W89
+            reason: "No need for VPC in this case"
+          - id: W92
+            reason: "No need for reserved concurrent executions"
+
+  RDSEOSLambda:
```
Comment (yprikhodko) [inline data-collection/deploy/module-rds-health.yaml]: This data is collected by reference module

File: data-collection/deploy/module-rds-health.yaml [hunk 6]
```diff
@@ -0,0 +1,1914 @@
...
+          BUCKET_NAME: !Ref DestinationBucket
+    Metadata:
+      cfn_nag:
+        rules_to_suppress:
+          - id: W89
+            reason: "No need for VPC in this case"
+          - id: W92
+            reason: "No need for reserved concurrent executions"
+
+  RDSMaintenanceLambda:
```
Comment (yprikhodko) [inline data-collection/deploy/module-rds-health.yaml]: This belongs to this module

File: data-collection/deploy/deploy-in-management-account.yaml [hunk 7]
```diff
@@ -153,6 +153,7 @@ Resources:
                   - !Sub "arn:${AWS::Partition}:iam::${DataCollectionAccountID}:role/${ResourcePrefix}RLS-LambdaRole"
                   - !Sub "arn:${AWS::Partition}:iam::${DataCollectionAccountID}:role/${ResourcePrefix}service-quotas-LambdaRole"
                   - !Sub "arn:${AWS::Partition}:iam::${DataCollectionAccountID}:role/${ResourcePrefix}identity_center-LambdaRole"
+                  - !Sub "arn:${AWS::Partition}:iam::${DataCollectionAccountID}:role/${ResourcePrefix}RHD-rds-LambdaRole"
```
Comment (yprikhodko) [inline data-collection/deploy/deploy-in-management-account.yaml]: I think it's not needed here as module-rds is LINKED-only and the RDS Lambda only ever assumes the multi-account role (role/${MultiAccountRoleName} on account *), never the management-account role. Payer-account RDS instances are covered by deploying deploy-in-linked-account.yaml into the management account (the account collector's list_accounts already returns the payer as an active member) — so nothing is needed in the management stack

File: data-collection/deploy/module-rds.yaml [hunk 8]
```diff
@@ -0,0 +1,753 @@
...
+                  return rds_client
+              except Exception as exc:
+                  logger.error(f"Error assuming role for account {account_id} in region {region}: {exc}")
+                  raise
+
+          def get_maintenance_actions(rds_client, region):
+              logger.info(f"Getting maintenance actions for region {region}")
+              try:
+                  response = rds_client.describe_pending_maintenance_actions()
+                  actions_count = len(response.get('PendingMaintenanceActions', []))
```
Comment (yprikhodko) [inline data-collection/deploy/module-rds.yaml]: SHouldn't we paginate here?

File: data-collection/deploy/module-rds.yaml
Comment (yprikhodko) [inline data-collection/deploy/module-rds.yaml:1]: Add the standard block to the Resources: section of module-rds.yaml (anywhere; next to the crawlers is fine):


  AnalyticsExecutor:
    Type: Custom::LambdaAnalyticsExecutor
    Properties:
      ServiceToken: !Ref LambdaAnalyticsARN
      Name: !Ref CFDataName

File: data-collection/deploy/module-rds.yaml [hunk 9]
```diff
@@ -0,0 +1,753 @@
...
+                  - 's3:ListBucket'
+                Resource:
+                  - !Sub
+                    - 'arn:${AWS::Partition}:s3:::${BucketName}'
+                    - BucketName: !Ref DestinationBucket
+                Condition:
+                  StringLike:
+                    s3:prefix:
+                      - 'rds-*'
+        - PolicyName: 'RDSMaintenanceDescribe'
```
Comment (yprikhodko) [inline data-collection/deploy/module-rds.yaml]: module-rds.yaml's LambdaRole carries an inline RDSMaintenanceDescribe policy (rds:DescribePendingMaintenanceActions + ec2:DescribeRegions), but the Lambda always assumes the multi-account role and issues the describe call with those creds — so these grants are never used (and ec2:DescribeRegions is never called). This also duplicates the authorization that correctly lives in RDSModulePolicy in deploy-in-linked-account.yaml. You can remove this policy from this stack

File: data-collection/deploy/module-rds.yaml [hunk 10]
```diff
@@ -0,0 +1,753 @@
...
+        Version: '2012-10-17'
+        Statement:
+          - Effect: 'Allow'
+            Principal:
+              Service: !Sub 'lambda.${AWS::URLSuffix}'
+            Action: 'sts:AssumeRole'
+      ManagedPolicyArns:
+        - !Sub 'arn:${AWS::Partition}:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole'
+      Path: /
+      Policies:
```
Comment (yprikhodko) [inline data-collection/deploy/module-rds.yaml:134]: The module declares and the parent passes DataBucketsKmsKeysArns, but LambdaRole has no KMS policy. Every sibling includes one:


- PolicyName: "KMS"
    ...
    Action:
      - "kms:GenerateDataKey"
    Resource: !Split [ ',', !Ref DataBucketsKmsKeysArns ]

File: data-collection/deploy/module-rds.yaml [hunk 11]
```diff
@@ -0,0 +1,753 @@
...
+        - PolicyName: 'S3Access'
```
Comment (yprikhodko) [inline data-collection/deploy/module-rds.yaml:135]: Sibling modules (module-rds-usage, module-transit-gateway, module-inventory) all grant a single S3 action — s3:PutObject on ${DestinationBucketARN}/* — because the collector Lambda only ever uploads. module-rds over-grants:


- PolicyName: 'S3Access'
    ...
    - Action:
        - 's3:PutObject'
        - 's3:GetObject'      # ← never used
    - Action:
        - 's3:ListBucket'     # ← never used (+ its rds-* prefix condition)
The Lambda code only calls s3client.upload_file(...) → PutObject. There's no get_object/download_file and no list_objects anywhere, so:

s3:GetObject — redundant, remove.
s3:ListBucket (and its StringLike s3:prefix: rds-* condition) — redundant, remove.

File: data-collection/deploy/module-rds.yaml [hunk 12]
```diff
@@ -0,0 +1,753 @@
...
+        Arn: !GetAtt StepFunctionmaintenance.Arn
+        RoleArn: !Ref SchedulerExecutionRoleARN
+        Input: !Sub |
+          {
+            "module": "${CFDataName}",
+            "name": "maintenance",
+            "params": "maintenance"
+          }
+
+  Crawlermaintenance:
```
Comment (yprikhodko) [inline data-collection/deploy/module-rds.yaml:483]: Align crawler configurations with other modules 
      Configuration: '{"Version":1.0,"Grouping":{"TableGroupingPolicy":"CombineCompatibleSchemas"},"CrawlerOutput":{"Partitions":{"AddOrUpdateBehavior":"InheritFromTable"},"Tables":{"TableThreshold":1,"AddOrUpdateBehavior":"MergeNewColumns"}}}'
      SchemaChangePolicy:
        UpdateBehavior: UPDATE_IN_DATABASE
        DeleteBehavior: LOG

File: data-collection/deploy/module-rds.yaml [hunk 13]
```diff
@@ -0,0 +1,753 @@
...
+        Arn: !GetAtt StepFunctionclustersnapshot.Arn
+        RoleArn: !Ref SchedulerExecutionRoleARN
+        Input: !Sub |
+          {
+            "module": "${CFDataName}",
+            "name": "clustersnapshot",
+            "params": "clustersnapshot"
+          }
+
+  Crawlerclustersnapshot:
```
Comment (yprikhodko) [inline data-collection/deploy/module-rds.yaml:602]: Align crawler configurations with other modules 
      Configuration: '{"Version":1.0,"Grouping":{"TableGroupingPolicy":"CombineCompatibleSchemas"},"CrawlerOutput":{"Partitions":{"AddOrUpdateBehavior":"InheritFromTable"},"Tables":{"TableThreshold":1,"AddOrUpdateBehavior":"MergeNewColumns"}}}'
      SchemaChangePolicy:
        UpdateBehavior: UPDATE_IN_DATABASE
        DeleteBehavior: LOG

File: data-collection/deploy/module-inventory.yaml [hunk 14]
```diff
@@ -315,21 +315,25 @@ Mappings:
             - Name: clonegroupid
               Type: string
             - Name: dbclustermembers
-              Type: string
+              Type: array<struct<DBInstanceIdentifier:string,IsClusterWriter:boolean,DBClusterParameterGroupStatus:string,PromotionTier:int>>
```
Comment (yprikhodko) [inline data-collection/deploy/module-inventory.yaml]: Prefer string over array<struct<…>> for dbclustermembers (and for taglist).

On main dbclustermembers is string; this PR changes it to a rigid array<struct<…>>. We should keep it as string. Glue crawlers infer nested JSON types from whatever data is present at crawl time, so an optional/variable field gets inferred as one type in a partition where it's populated and a different type where it's empty — the partitions then have mismatched schemas and fail to load in the Athena view. We already hit exactly this with event_metadata in the Health Events module . A string column is immune: a JSON blob is always a string regardless of content, so it never drifts and it survives AWS API schema changes.

Query it with JSON functions in the view instead of baking the shape into the table, e.g.:


SELECT
  c.dbclusteridentifier,
  m.dbinstanceidentifier,
  m.isclusterwriter,
  m.promotiontier
FROM inventory_rds_db_clusters_data c
CROSS JOIN UNNEST(
  CAST(json_parse(c.dbclustermembers) AS ARRAY(ROW(
    dbinstanceidentifier varchar,
    isclusterwriter boolean,
    promotiontier int
  )))
) AS t(m);
This keeps the raw collection stable (string) and moves the structure into a derived Athena view, per the module guidelines. taglist has the same empty-most-of-the-time risk, so I'd lean string + json_extract/UNNEST there too rather than array<struct<Key,Value>>

File: data-collection/deploy/deploy-in-management-account.yaml (hunk 7 above)
Comment (anishkgemr) [reply data-collection/deploy/deploy-in-management-account.yaml]: Thank you, removed. module-rds is LINKED-only: its Lambda only assumes the multi-account role, never the
management-account role, and payer-account RDS is covered by deploying deploy-in-linked-account.yaml into the management account. Dropped the stale principal line. (599ca7b)

File: data-collection/deploy/module-rds.yaml (hunk 8 above)
Comment (anishkgemr) [reply data-collection/deploy/module-rds.yaml]: Thank you, switched to the boto3 paginator (get_paginator('describe_pending_maintenance_actions')) so all pages are collected; the single call could truncate at ~100 records for large accounts. (7818d46)

File: data-collection/deploy/module-rds.yaml
Comment (anishkgemr) [reply data-collection/deploy/module-rds.yaml:1]: Added the standard AnalyticsExecutor (Custom::LambdaAnalyticsExecutor) next to the crawlers, wired to
LambdaAnalyticsARN / CFDataName. (7818d46)

File: data-collection/deploy/module-rds.yaml (hunk 9 above)
Comment (anishkgemr) [reply data-collection/deploy/module-rds.yaml]: Removed. The Lambda always assumes the multi-account role and issues the describe with those creds, so the
execution-role grant was never exercised — and ec2:DescribeRegions is never called (regions come from RegionsInScope). The real authorization already lives in RDSModulePolicy in deploy-in-linked-account.yaml. Also removed the now-orphaned W11 suppression. (7818d46)

File: data-collection/deploy/module-rds.yaml (hunk 10 above)
Comment (anishkgemr) [reply data-collection/deploy/module-rds.yaml:134]: Added — introduced a NeedDataBucketsKms condition and Matches module-inventory/module-backup. (7818d46)

File: data-collection/deploy/module-rds.yaml (hunk 11 above)
Comment (anishkgemr) [reply data-collection/deploy/module-rds.yaml:135]: Trimmed to a single s3:PutObject on ${DestinationBucketARN}/*. GetObject/ListBucket were leftovers from an earlier design; removed. Matches module-rds-usage/transit-gateway. (7818d46)

File: data-collection/deploy/module-rds.yaml (hunk 12 above)
Comment (anishkgemr) [reply data-collection/deploy/module-rds.yaml:483]: Aligned with the standard schema-safe config (Partitions.InheritFromTable + Tables.MergeNewColumns) and added SchemaChangePolicy (UPDATE_IN_DATABASE / DeleteBehavior: LOG). (7818d46)

File: data-collection/deploy/module-rds.yaml (hunk 13 above)
Comment (anishkgemr) [reply data-collection/deploy/module-rds.yaml:602]: Same alignment applied here as well — both crawlers now use the standard config + SchemaChangePolicy. (7818d46)

File: data-collection/deploy/module-inventory.yaml (hunk 14 above)
Comment (anishkgemr) [reply data-collection/deploy/module-inventory.yaml]: Done — reverted dbclustermembers and taglist to string and parse them at query time in the derived views via CAST(json_parse(...) AS ARRAY(ROW(...))) + UNNEST, per the guidance. Validated in Athena including the empty [] case and against a real Aurora cluster. Data-collection: 2f4206c; view changes in the dashboard PR#1521.
